src/main/java/fi/aalto/cs/apluscourses/presentation/exercise/ExerciseGroupViewModel.java

```diff
@@ -45,4 +46,61 @@ public String getPresentableName() {
   public List<ExerciseViewModel> getSubtrees() {
     return getExerciseViewModels();
   }
+
+  // O1_SPECIFIC
+  private static final Comparator<ExerciseViewModel> EXERCISE_COMPARATOR = (exercise1, exercise2)
+      -> {
+    int chapter1 = getExerciseChapter(exercise1);
+    int chapter2 = getExerciseChapter(exercise2);
+    if (chapter1 > chapter2) {
+      return 1;
+    }
+    if (chapter2 > chapter1) {
+      return -1;
+    }
+
+    // Feedback exercises are always last
+    if ("Feedback".equals(exercise1.getPresentableName())) {
+      return 1;
+    }
+    if ("Feedback".equals(exercise2.getPresentableName())) {
+      return -1;
+    }
+
+    int number1 = getExerciseNumber(exercise1);
+    int number2 = getExerciseNumber(exercise2);
+    if (number1 > number2) {
+      return 1;
+    }
+    if (number2 > number1) {
+      return -1;
+    }
+    return 0;
+  };
+
+  private static final Pattern CHAPTER_REGEX = Pattern.compile("\\/w[0-9]+\\/ch([0-9]+)\\/");
+
+  private static int getExerciseChapter(@NotNull ExerciseViewModel exercise) {
+    String htmlUrl = exercise.getModel().getHtmlUrl();
+    Matcher matcher = CHAPTER_REGEX.matcher(htmlUrl);
+    matcher.find();
+    try {
+      return Integer.parseInt(matcher.group(1));
+    } catch (IllegalStateException | NumberFormatException | IndexOutOfBoundsException e) {
+      return -1;
+    }
+  }
+
+  private static final Pattern NUMBER_REGEX = Pattern.compile("w[0-9]+_ch[0-9]+_([0-9]+)");
+
+  private static int getExerciseNumber(@NotNull ExerciseViewModel exercise) {
```

Comment:
could `getExerciseNumber` and `getExerciseChapter` be a single method taking additionally a `Patter` regex used for matching?

Reply:
Hmm true there is some unnecesssary duplication there

Then again I would prefer to burn this code with a flamethrower :smile: